File: internal/importer/government_funding_test.go
package importer

import (
	"context"
	"os"
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"gorm.io/driver/sqlite"
	"gorm.io/gorm"
	"gorm.io/gorm/logger"

	"github.com/eenemeene/kitamanager-go/internal/models"
	"github.com/eenemeene/kitamanager-go/internal/store"
)

func setupTestDB(t *testing.T) *gorm.DB {
	db, err := gorm.Open(sqlite.Open(":memory:"), &gorm.Config{
		Logger: logger.Default.LogMode(logger.Silent),
	})
	require.NoError(t, err)

	err = db.AutoMigrate(
		&models.GovernmentFunding{},
		&models.GovernmentFundingPeriod{},
		&models.GovernmentFundingProperty{},
	)
	require.NoError(t, err)

	return db
}

func createTestYAMLFile(t *testing.T, content string) string {
	tmpDir := t.TempDir()
	filePath := filepath.Join(tmpDir, "test-government-funding.yaml")
	err := os.WriteFile(filePath, []byte(content), 0644)
	require.NoError(t, err)
	return filePath
}

func TestEuroToCents(t *testing.T) {
	tests := []struct {
		name     string
		eur      float64
		expected int
	}{
		{"zero", 0, 0},
		{"whole number", 100.00, 10000},
		{"with cents", 1668.47, 166847},
		{"small amount", 0.01, 1},
		{"rounding up", 100.005, 10001},
		{"rounding down", 100.004, 10000},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := euroToCents(tt.eur)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestParseDate(t *testing.T) {
	tests := []struct {
		name        string
		input       string
		expectError bool
		expectYear  int
		expectMonth int
		expectDay   int
	}{
		{"valid date", "2023-03-01", false, 2023, 3, 1},
		{"valid date 2", "2022-01-01", false, 2022, 1, 1},
		{"invalid format", "01-03-2023", true, 0, 0, 0},
		{"empty string", "", true, 0, 0, 0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := parseDate(tt.input)
			if tt.expectError {
				assert.Error(t, err)
			} else {
				require.NoError(t, err)
				assert.Equal(t, tt.expectYear, result.Year())
				assert.Equal(t, tt.expectMonth, int(result.Month()))
				assert.Equal(t, tt.expectDay, result.Day())
			}
		})
	}
}

func TestImportGovernmentFundingFromFile(t *testing.T) {
	yamlContent := `---
-
  from: '2023-03-01'
  to: ''
  entries:
    - age: [0,2]
      properties:
        - key: care_type
          value: ganztag
          payment: 1668.47
          requirement: 0.261
        - key: care_type
          value: halbtag
          payment: 1066.64
          requirement: 0.14
-
  from: '2022-01-01'
  to: '2023-03-01'
  comment: |
    Test period
  entries:
    - age: [0,2]
      properties:
        - key: care_type
          value: ganztag
          payment: 1640.43
          requirement: 0.261
`

	db := setupTestDB(t)
	fundingStore := store.NewGovernmentFundingStore(db)
	importer := NewGovernmentFundingImporter(db, fundingStore)

	filePath := createTestYAMLFile(t, yamlContent)

	// First import should succeed
	fundingID, err := importer.ImportGovernmentFundingFromFile(context.Background(), filePath, "berlin")
	require.NoError(t, err)
	assert.NotZero(t, fundingID)

	// Verify government funding was created
	funding, err := fundingStore.FindByIDWithDetails(context.Background(), fundingID, 0, nil)
	require.NoError(t, err)
	assert.Equal(t, "Berlin Kita-Förderung", funding.Name)
	assert.Equal(t, "berlin", funding.State)
	assert.Len(t, funding.Periods, 2)

	// Periods are ordered by from_date DESC, so the latest (2023-03-01) is first
	period1 := funding.Periods[0] // 2023-03-01 onwards (ongoing - no end date)
	assert.Nil(t, period1.To)
	assert.Len(t, period1.Properties, 2)

	// Check ganztag property has correct cents conversion and age range
	var ganztag *models.GovernmentFundingProperty
	for i := range period1.Properties {
		if period1.Properties[i].Key == "care_type" && period1.Properties[i].Value == "ganztag" {
			ganztag = &period1.Properties[i]
			break
		}
	}
	require.NotNil(t, ganztag)
	assert.Equal(t, 166847, ganztag.Payment) // 1668.47 EUR = 166847 cents
	assert.Equal(t, 0.261, ganztag.Requirement)
	require.NotNil(t, ganztag.MinAge)
	require.NotNil(t, ganztag.MaxAge)
	assert.Equal(t, 0, *ganztag.MinAge)
	// Both MinAge and MaxAge are inclusive: [0,2] means ages 0, 1, and 2
	assert.Equal(t, 2, *ganztag.MaxAge)

	// Check second period has end date and comment
	period2 := funding.Periods[1] // 2022-01-01 to 2023-03-01
	require.NotNil(t, period2.To)
	assert.Equal(t, 2023, period2.To.Year())
	assert.Contains(t, period2.Comment, "Test period")
}

func TestImportGovernmentFundingFromFile_Idempotency(t *testing.T) {
	yamlContent := `---
-
  from: '2023-03-01'
  to: ''
  entries:
    - age: [0,2]
      properties:
        - key: care_type
          value: ganztag
          payment: 1668.47
          requirement: 0.261
`

	db := setupTestDB(t)
	fundingStore := store.NewGovernmentFundingStore(db)
	importer := NewGovernmentFundingImporter(db, fundingStore)

	filePath := createTestYAMLFile(t, yamlContent)

	// First import
	fundingID1, err := importer.ImportGovernmentFundingFromFile(context.Background(), filePath, "berlin")
	require.NoError(t, err)

	// Second import should return ErrGovernmentFundingExists
	fundingID2, err := importer.ImportGovernmentFundingFromFile(context.Background(), filePath, "berlin")
	assert.ErrorIs(t, err, ErrGovernmentFundingExists)
	assert.Equal(t, fundingID1, fundingID2)

	// Verify only one government funding exists
	fundings, total, err := fundingStore.FindAll(context.Background(), 100, 0)
	require.NoError(t, err)
	assert.Equal(t, int64(1), total)
	assert.Len(t, fundings, 1)
}

func TestImportGovernmentFundingFromFile_FarFutureDateTreatedAsOngoing(t *testing.T) {
	yamlContent := `---
-
  from: '2023-03-01'
  to: '2060-01-01'
  entries:
    - age: [0,2]
      properties:
        - key: care_type
          value: ganztag
          payment: 1668.47
          requirement: 0.261
`

	db := setupTestDB(t)
	fundingStore := store.NewGovernmentFundingStore(db)
	importer := NewGovernmentFundingImporter(db, fundingStore)

	filePath := createTestYAMLFile(t, yamlContent)

	fundingID, err := importer.ImportGovernmentFundingFromFile(context.Background(), filePath, "berlin")
	require.NoError(t, err)

	funding, err := fundingStore.FindByIDWithDetails(context.Background(), fundingID, 0, nil)
	require.NoError(t, err)
	require.Len(t, funding.Periods, 1)

	// 2060-01-01 should be treated as nil (ongoing)
	assert.Nil(t, funding.Periods[0].To)
}

func TestImportGovernmentFundingFromFile_FileNotFound(t *testing.T) {
	db := setupTestDB(t)
	fundingStore := store.NewGovernmentFundingStore(db)
	importer := NewGovernmentFundingImporter(db, fundingStore)

	_, err := importer.ImportGovernmentFundingFromFile(context.Background(), "/nonexistent/file.yaml", "berlin")
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "failed to read file")
}

func TestImportGovernmentFundingFromFile_InvalidYAML(t *testing.T) {
	db := setupTestDB(t)
	fundingStore := store.NewGovernmentFundingStore(db)
	importer := NewGovernmentFundingImporter(db, fundingStore)

	filePath := createTestYAMLFile(t, "invalid: yaml: content: [")

	_, err := importer.ImportGovernmentFundingFromFile(context.Background(), filePath, "berlin")
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "failed to parse YAML")
}

func TestImportGovernmentFundingFromFile_InvalidState(t *testing.T) {
	db := setupTestDB(t)
	fundingStore := store.NewGovernmentFundingStore(db)
	importer := NewGovernmentFundingImporter(db, fundingStore)

	_, err := importer.ImportGovernmentFundingFromFile(context.Background(), "/any/path.yaml", "invalid")
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "invalid state")
}

File: frontend/src/lib/schemas/__tests__/government-funding.test.ts
import {
  governmentFundingSchema,
  governmentFundingPeriodSchema,
  governmentFundingPropertySchema,
} from '../government-funding';

describe('governmentFundingSchema', () => {
  it('accepts valid funding', () => {
    expect(
      governmentFundingSchema.safeParse({ name: 'Berlin Funding', state: 'berlin' }).success
    ).toBe(true);
  });

  it('rejects empty name', () => {
    expect(governmentFundingSchema.safeParse({ name: '', state: 'berlin' }).success).toBe(false);
  });

  it('rejects empty state', () => {
    expect(governmentFundingSchema.safeParse({ name: 'Test', state: '' }).success).toBe(false);
  });
});

describe('governmentFundingPeriodSchema', () => {
  const valid = {
    from: '2024-01-01',
    full_time_weekly_hours: 39,
  };

  it('accepts valid period without end date', () => {
    expect(governmentFundingPeriodSchema.safeParse(valid).success).toBe(true);
  });

  it('accepts valid period with end date after start', () => {
    expect(governmentFundingPeriodSchema.safeParse({ ...valid, to: '2024-12-31' }).success).toBe(
      true
    );
  });

  it('rejects end date before start date', () => {
    const result = governmentFundingPeriodSchema.safeParse({ ...valid, to: '2023-06-01' });
    expect(result.success).toBe(false);
    if (!result.success) {
      expect(result.error.issues[0].path).toContain('to');
    }
  });

  it('rejects weekly hours below 0.1', () => {
    expect(
      governmentFundingPeriodSchema.safeParse({ ...valid, full_time_weekly_hours: 0 }).success
    ).toBe(false);
  });

  it('rejects weekly hours above 80', () => {
    expect(
      governmentFundingPeriodSchema.safeParse({ ...valid, full_time_weekly_hours: 81 }).success
    ).toBe(false);
  });

  it('accepts optional comment', () => {
    const result = governmentFundingPeriodSchema.safeParse({
      ...valid,
      comment: 'Test comment',
    });
    expect(result.success).toBe(true);
  });
});

describe('governmentFundingPropertySchema', () => {
  const valid = {
    key: 'care_type',
    value: 'ganztag',
    label: 'Ganztag',
    payment_euros: 1668.47,
    requirement: 0.261,
  };

  it('accepts valid property', () => {
    expect(governmentFundingPropertySchema.safeParse(valid).success).toBe(true);
  });

  it('accepts property with age range', () => {
    expect(
      governmentFundingPropertySchema.safeParse({ ...valid, min_age: 0, max_age: 2 }).success
    ).toBe(true);
  });

  it('accepts null age fields', () => {
    expect(
      governmentFundingPropertySchema.safeParse({ ...valid, min_age: null, max_age: null }).success
    ).toBe(true);
  });

  it('rejects negative payment', () => {
    expect(governmentFundingPropertySchema.safeParse({ ...valid, payment_euros: -1 }).success).toBe(
      false
    );
  });

  it('rejects negative requirement', () => {
    expect(governmentFundingPropertySchema.safeParse({ ...valid, requirement: -0.1 }).success).toBe(
      false
    );
  });

  it('rejects empty key', () => {
    expect(governmentFundingPropertySchema.safeParse({ ...valid, key: '' }).success).toBe(false);
  });

  it('rejects empty value', () => {
    expect(governmentFundingPropertySchema.safeParse({ ...valid, value: '' }).success).toBe(false);
  });
});
